cmd/kubeadm/app/phases/upgrade/prepull.go

```diff
@@ -0,0 +1,177 @@
+		fmt.Printf("[upgrade/prepull] WARNING: Unable to cleanup prepull DaemonSet %s\n", component)
+	}
+	fmt.Printf("[upgrade/prepull] Prepulled image for component %s.\n", component)
+}
+
+// PrepullImagesInParallell creates DaemonSets synchronously but waits in parallell for the images to pull
+func PrepullImagesInParallell(kubePrepuller prepuller, timeout time.Duration) error {
+	componentsToPrepull := constants.MasterComponents
+	fmt.Printf("[upgrade/prepull] Will prepull images for components %v\n", componentsToPrepull)
+
+	timeoutChan := time.After(timeout)
+
+	// Synchronously create the DaemonSets
+	for _, component := range componentsToPrepull {
+		kubePrepuller.CreateFunc(component)
+	}
+
+	// Create a channel for streaming data from goroutines that run in parallell to a blocking for loop that cleans up
+	prePulledChan := make(chan string, len(componentsToPrepull))
+	for _, component := range componentsToPrepull {
+		go func(c string) {
+			// Wait as long as needed. This WaitFunc call should be blocking until completetion
+			kubePrepuller.WaitFunc(c)
+			// When the task is done, go ahead and cleanup by sending the name to the channel
+			prePulledChan <- c
+		}(component)
+	}
+
+	// This call blocks until all expected messages are received from the channel or errors out if timeoutChan fires.
+	// For every successful wait, kubePrepuller.DeleteFunc is executed
+	if err := waitForItemsFromChan(timeoutChan, prePulledChan, len(componentsToPrepull), kubePrepuller.DeleteFunc); err != nil {
+		return err
+	}
+
+	fmt.Println("[upgrade/prepull] Successfully prepulled the images for all the control plane components")
+	return nil
+}
+
+// waitForItemsFromChan waits for n elements from stringChan with a timeout. For every item received from stringChan, cleanupFunc is executed
+func waitForItemsFromChan(timeoutChan <-chan time.Time, stringChan chan string, n int, cleanupFunc func(string)) error {
+	i := 0
+	for {
+		select {
+		case <-timeoutChan:
+			return fmt.Errorf("The prepull operation timed out")
+		case result := <-stringChan:
+			i++
+			cleanupFunc(result)
+			if i == n {
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+func addPrepullPrefix(component string) string {
+	return fmt.Sprintf("%s%s", prepullPrefix, component)
+}
+
+func buildPrePullDaemonSet(cfg *kubeadmapi.MasterConfiguration, component string) *extensions.DaemonSet {
+	var gracePeriodSecs int64 = 0
+	return &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{Name: addPrepullPrefix(component)},
+		Spec: extensions.DaemonSetSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"k8s-app": addPrepullPrefix(component),
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:    component,
+							Image:   images.GetCoreImage(component, cfg.ImageRepository, cfg.KubernetesVersion, cfg.UnifiedControlPlaneImage),
+							Command: []string{"sleep", "3600"},
```

Comment:
Initially I was wondering whether 60s would be enough for user's on a really slow box (hyperkube images can be over 600MB) but I guess the container command doesn't affect the kubelet's image pull, right?